src/middlewared/middlewared/plugins/smb.py

```diff
@@ -72,6 +72,130 @@ def key_cp(encoding):
+        take ownership of Files, override ACLs, view and modify user quotas, and administer
+        the server via the Computer Management MMC Snap-In. Unfortuntely, group membership
+        must be managed via "net groupmap listmem|addmem|delmem", which requires that
+        winbind be running when the commands are executed. In this situation net command
+        will fail with WBC_ERR_WINBIND_NOT_AVAILABLE. If this error message is returned, then
+        flag for a deferred command retry when service starts.
+
+        @param-in (admin_group): This is the group to add to BUILTIN\\Administrators. If unset, then
+            look up the value in the config db.
+        @param-in (check_deferred): If this is True, then only perform the group mapping if this has
+            been flagged as in need of deferred setup (i.e. samba wasn't running when it was initially
+            called). This is to avoid unecessarily calling during service start.
+        """
+        ret = {
+            'status': False,
+            'message': None,
+        }
+        if check_deferred:
+            is_deferred = await self.middleware.call('cache.has_key', 'SMB_SET_ADMIN')
+            if not is_deferred:
+                return {'status': True, 'message': 'No delayed action detected. Exiting.'}
+            else:
+                await self.middleware.call('cache.pop', 'SMB_SET_ADMIN')
+
+        if not admin_group:
+            smb = await self.middleware.call('smb.config')
+            admin_group = smb['admin_group']
+
+        # We must use GIDs because wbinfo --name-to-sid expects a domain prefix "FREENAS\user"
+        group = await self.middleware.call("notifier.get_group_object", admin_group)
+        if not group:
+            return {'status': False, 'message': f"Failed to look up group {admin_group}"}
+
+        sid = await self.wbinfo_gidtosid(group[2])
+        if not sid['status']:
+            self.logger.debug(f"Failed to convert {admin_group} to SID: ({sid['message']})")
+            if "WBC_ERR_WINBIND_NOT_AVAILABLE" in sid['message']:
+                self.logger.debug("Delaying admin group add until winbind starts")
+                await self.middleware.call('cache.put', 'SMB_SET_ADMIN', True)
+
+                return {'status': True, 'message': 'Delaying action until winbind starts'}
+
+            return sid
+        ret = await self.validate_admin_groups(sid['message'])
+        if not ret['status'] or ret['message'] == 'SID_IS_PRESENT':
+            return ret
+
+        proc = await Popen(
+            ['/usr/local/bin/net', 'groupmap', 'addmem', 'S-1-5-32-544', sid['message']],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        output = (await proc.communicate())
+        if output[1]:
+            return {'status': False, 'message': {output[1].decode()}}
+
+        return {'status': True, 'message': f"Successfully added {admin_group} to BUILTIN\\Administrators"}
```

Comment:
The reason to change to exceptions is also that now messages like this go to nowhere (and they should go to logging)